File: projectforge-business/src/main/java/org/projectforge/business/timesheet/TimesheetExport.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.timesheet;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;

import org.apache.poi.hssf.util.HSSFColor;
import org.projectforge.business.common.OutputType;
import org.projectforge.business.excel.CellFormat;
import org.projectforge.business.excel.ContentProvider;
import org.projectforge.business.excel.ExportCell;
import org.projectforge.business.excel.ExportColumn;
import org.projectforge.business.excel.ExportRow;
import org.projectforge.business.excel.ExportSheet;
import org.projectforge.business.excel.ExportWorkbook;
import org.projectforge.business.excel.I18nExportColumn;
import org.projectforge.business.excel.PropertyMapping;
import org.projectforge.business.fibu.KundeDO;
import org.projectforge.business.fibu.ProjektDO;
import org.projectforge.business.fibu.kost.Kost2DO;
import org.projectforge.business.multitenancy.TenantRegistryMap;
import org.projectforge.business.task.TaskNode;
import org.projectforge.business.task.TaskTree;
import org.projectforge.business.task.formatter.TaskFormatter;
import org.projectforge.business.tasktree.TaskTreeHelper;
import org.projectforge.business.user.UserGroupCache;
import org.projectforge.export.MyXlsContentProvider;
import org.projectforge.framework.persistence.user.api.ThreadLocalUserContext;
import org.projectforge.framework.persistence.user.entities.PFUserDO;
import org.projectforge.framework.time.DateFormatType;
import org.projectforge.framework.time.DateFormats;
import org.projectforge.framework.time.DateHolder;
import org.projectforge.framework.time.DateTimeFormatter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

/**
 * For excel export.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
@Service
public class TimesheetExport
{

  private static final org.apache.log4j.Logger log = org.apache.log4j.Logger.getLogger(TimesheetExport.class);

  @Autowired
  private DateTimeFormatter dateTimeFormatter;

  private class MyContentProvider extends MyXlsContentProvider
  {
    public MyContentProvider(final ExportWorkbook workbook)
    {
      super(workbook);
    }

    @Override
    public MyContentProvider updateRowStyle(final ExportRow row)
    {
      for (final ExportCell cell : row.getCells()) {
        final CellFormat format = cell.ensureAndGetCellFormat();
        format.setFillForegroundColor(HSSFColor.WHITE.index);
        switch (row.getRowNum()) {
          case 0:
            format.setFont(FONT_HEADER);
            break;
          case 1:
            format.setFont(FONT_NORMAL_BOLD);
            // alignment = CellStyle.ALIGN_CENTER;
            break;
          default:
            format.setFont(FONT_NORMAL);
            if (row.getRowNum() % 2 == 0) {
              format.setFillForegroundColor(HSSFColor.GREY_25_PERCENT.index);
            }
            break;
        }
      }
      return this;
    }

    @Override
    public ContentProvider newInstance()
    {
      return new MyContentProvider(this.workbook);
    }
  }

  ;

  private enum Col
  {
    USER, KUNDE, PROJEKT, KOST2, WEEK_OF_YEAR, DAY_OF_WEEK, START_TIME, STOP_TIME, DURATION, HOURS, LOCATION, TASK_TITLE, REFERENCE, SHORT_DESCRIPTION, DESCRIPTION, TASK_PATH, ID;
  }

  /**
   * Exports the filtered list as table with almost all fields.
   */
  public byte[] export(final List<TimesheetDO> list)
  {
    log.info("Exporting timesheet list.");
    final ExportWorkbook xls = new ExportWorkbook();
    final ContentProvider contentProvider = new MyContentProvider(xls);
    // create a default Date format and currency column
    xls.setContentProvider(contentProvider);

    final String sheetTitle = ThreadLocalUserContext.getLocalizedString("timesheet.timesheets");
    final ExportSheet sheet = xls.addSheet(sheetTitle);
    sheet.createFreezePane(8, 1);

    final ExportColumn[] cols = new ExportColumn[] { //
        new I18nExportColumn(Col.USER, "timesheet.user", MyXlsContentProvider.LENGTH_USER),
        new I18nExportColumn(Col.KUNDE, "fibu.kunde", MyXlsContentProvider.LENGTH_STD),
        new I18nExportColumn(Col.PROJEKT, "fibu.projekt", MyXlsContentProvider.LENGTH_STD),
        new I18nExportColumn(Col.KOST2, "fibu.kost2", MyXlsContentProvider.LENGTH_KOSTENTRAEGER),
        new I18nExportColumn(Col.WEEK_OF_YEAR, "calendar.weekOfYearShortLabel", 4),
        new I18nExportColumn(Col.DAY_OF_WEEK, "calendar.dayOfWeekShortLabel", 4),
        new I18nExportColumn(Col.START_TIME, "timesheet.startTime", MyXlsContentProvider.LENGTH_DATETIME),
        new I18nExportColumn(Col.STOP_TIME, "timesheet.stopTime", MyXlsContentProvider.LENGTH_TIMESTAMP),
        new I18nExportColumn(Col.DURATION, "timesheet.duration", MyXlsContentProvider.LENGTH_DURATION),
        new I18nExportColumn(Col.HOURS, "hours", MyXlsContentProvider.LENGTH_DURATION),
        new I18nExportColumn(Col.LOCATION, "timesheet.location", MyXlsContentProvider.LENGTH_STD),
        new I18nExportColumn(Col.TASK_TITLE, "task.title", MyXlsContentProvider.LENGTH_STD),
        new I18nExportColumn(Col.REFERENCE, "task.reference", MyXlsContentProvider.LENGTH_STD),
        new I18nExportColumn(Col.SHORT_DESCRIPTION, "shortDescription", MyXlsContentProvider.LENGTH_STD),
        new I18nExportColumn(Col.DESCRIPTION, "timesheet.description", MyXlsContentProvider.LENGTH_EXTRA_LONG),
        new I18nExportColumn(Col.TASK_PATH, "task.path", MyXlsContentProvider.LENGTH_EXTRA_LONG),
        new I18nExportColumn(Col.ID, "id", MyXlsContentProvider.LENGTH_ID) };

    // column property names
    sheet.setColumns(cols);

    final ContentProvider sheetProvider = sheet.getContentProvider();
    sheetProvider.putFormat(Col.START_TIME, "yyyy-MM-dd HH:mm");
    sheetProvider.putFormat(Col.STOP_TIME, "HH:mm");
    sheetProvider.putFormat(Col.DURATION, "[h]:mm");
    sheetProvider.putFormat(Col.HOURS, "#,##0.00");
    sheetProvider.putFormat(Col.ID, "0");

    final PropertyMapping mapping = new PropertyMapping();
    final TaskTree taskTree = TaskTreeHelper.getTaskTree();
    final UserGroupCache userGroupCache = TenantRegistryMap.getInstance().getTenantRegistry().getUserGroupCache();
    for (final TimesheetDO timesheet : list) {
      final TaskNode node = taskTree.getTaskNodeById(timesheet.getTaskId());
      final PFUserDO user = userGroupCache.getUser(timesheet.getUserId());
      mapping.add(Col.USER, user.getFullname());
      final Kost2DO kost2 = timesheet.getKost2();
      String kost2Name = null;
      String projektName = null;
      String kundeName = null;
      if (kost2 != null) {
        kost2Name = kost2.getShortDisplayName();
        final ProjektDO projekt = kost2.getProjekt();
        if (projekt != null) {
          projektName = projekt.getName();
          final KundeDO kunde = projekt.getKunde();
          if (kunde != null) {
            kundeName = kunde.getName();
          } else {
          }
        }
      }
      mapping.add(Col.KOST2, kost2Name);
      mapping.add(Col.PROJEKT, projektName);
      mapping.add(Col.KUNDE, kundeName);
      mapping.add(Col.TASK_TITLE, node.getTask().getTitle());
      mapping.add(Col.TASK_PATH, TaskFormatter.getTaskPath(timesheet.getTaskId(), null, true, OutputType.PLAIN));
      mapping.add(Col.WEEK_OF_YEAR, timesheet.getFormattedWeekOfYear());
      mapping.add(Col.DAY_OF_WEEK, dateTimeFormatter.getFormattedDate(timesheet.getStartTime(), DateFormats
          .getFormatString(DateFormatType.DAY_OF_WEEK_SHORT)));
      final DateHolder startTime = new DateHolder(timesheet.getStartTime());
      final DateHolder stopTime = new DateHolder(timesheet.getStopTime());
      mapping.add(Col.START_TIME, startTime);
      mapping.add(Col.STOP_TIME, stopTime);
      final BigDecimal seconds = new BigDecimal(timesheet.getDuration() / 1000); // Seconds
      final BigDecimal duration = seconds.divide(new BigDecimal(60 * 60 * 24), 8, RoundingMode.HALF_UP); // Fraction of day (24 hours)
      mapping.add(Col.DURATION, duration.doubleValue());
      final BigDecimal hours = seconds.divide(new BigDecimal(60 * 60), 2, RoundingMode.HALF_UP);
      mapping.add(Col.HOURS, hours.doubleValue());
      mapping.add(Col.LOCATION, timesheet.getLocation());
      mapping.add(Col.REFERENCE, node.getReference());
      mapping.add(Col.SHORT_DESCRIPTION, timesheet.getShortDescription());
      mapping.add(Col.DESCRIPTION, timesheet.getDescription());
      mapping.add(Col.ID, timesheet.getId());
      sheet.addRow(mapping.getMapping(), 0);
    }
    sheet.setZoom(3, 4); // 75%

    return xls.getAsByteArray();
  }

  public void setDateTimeFormatter(final DateTimeFormatter dateTimeFormatter)
  {
    this.dateTimeFormatter = dateTimeFormatter;
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/business/timesheet/TimesheetExport.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2023 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.timesheet

import de.micromata.merlin.excel.ExcelWorkbook
import mu.KotlinLogging
import org.projectforge.business.common.OutputType
import org.projectforge.business.task.TaskFormatter.Companion.getTaskPath
import org.projectforge.business.task.TaskTree
import org.projectforge.business.user.UserGroupCache
import org.projectforge.common.DateFormatType
import org.projectforge.excel.ExcelUtils
import org.projectforge.framework.i18n.translate
import org.projectforge.framework.persistence.user.api.ThreadLocalUserContext
import org.projectforge.framework.time.DateFormats.getFormatString
import org.projectforge.framework.time.DateTimeFormatter
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.math.RoundingMode

private val log = KotlinLogging.logger {}

/**
 * For excel export.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
@Service
open class TimesheetExport {
  @Autowired
  private lateinit var dateTimeFormatter: DateTimeFormatter

  @Autowired
  private lateinit var taskTree: TaskTree

  @Autowired
  private lateinit var userGroupCache: UserGroupCache

  /**
   * Exports the filtered list as table with almost all fields.
   */
  open fun export(list: List<TimesheetDO>): ByteArray {
    log.info("Exporting timesheet list.")
    ExcelWorkbook.createEmptyWorkbook(ThreadLocalUserContext.locale!!).use { workbook ->
      val sheet = workbook.createOrGetSheet(translate("timesheet.timesheets"))
      val boldFont = workbook.createOrGetFont("bold", bold = true)
      val boldStyle = workbook.createOrGetCellStyle("hr", font = boldFont)
      val wrapTextStyle = workbook.createOrGetCellStyle("wrap")
      val timeFormat = workbook.ensureDateCellStyle("HH:mm")
      val durationFormat = workbook.ensureDateCellStyle("[h]:mm")
      val hoursFormat = workbook.createOrGetCellStyle("hours")
      hoursFormat.dataFormat = workbook.createDataFormat().getFormat("#,##0.00");
      val idFormat = workbook.createOrGetCellStyle("id")
      idFormat.dataFormat = workbook.createDataFormat().getFormat("0");

      wrapTextStyle.wrapText = true
      sheet.registerColumn(translate("timesheet.user"), "user").withSize(ExcelUtils.Size.USER)
      sheet.registerColumn(translate("fibu.kunde"), "kunde").withSize(ExcelUtils.Size.STANDARD)
      sheet.registerColumn(translate("fibu.projekt"), "projekt").withSize(ExcelUtils.Size.STANDARD)
      sheet.registerColumn(translate("fibu.kost2"), "kost2").withSize(ExcelUtils.Size.KOSTENTRAEGER)
      sheet.registerColumn(translate("calendar.weekOfYearShortLabel"), "weekOfYearShortLabel").withSize(4)
      sheet.registerColumn(translate("calendar.dayOfWeekShortLabel"), "dayOfWeekShortLabel").withSize(4)
      ExcelUtils.registerColumn(sheet, TimesheetDO::startTime, ExcelUtils.Size.TIMESTAMP)
      val stopTimeColDef = ExcelUtils.registerColumn(sheet, TimesheetDO::stopTime, 8)
      sheet.registerColumn(translate("timesheet.duration"), "duration").withSize(ExcelUtils.Size.DURATION)
      sheet.registerColumn(translate("hours"), "hours").withSize(ExcelUtils.Size.DURATION)
      ExcelUtils.registerColumn(sheet, TimesheetDO::location)
      ExcelUtils.registerColumn(sheet, TimesheetDO::reference)
      sheet.registerColumn(translate("task"), "task.title").withSize(ExcelUtils.Size.STANDARD)
      sheet.registerColumn(translate("timesheet.taskReference"), "taskReference").withSize(ExcelUtils.Size.STANDARD)
      sheet.registerColumn(translate("shortDescription"), "shortDescription").withSize(ExcelUtils.Size.EXTRA_LONG)
      val descriptionColDef = ExcelUtils.registerColumn(sheet, TimesheetDO::description).withSize(ExcelUtils.Size.EXTRA_LONG)
      sheet.registerColumn(translate("task.path"), "taskPath").withSize(ExcelUtils.Size.TASK_PATH)
      sheet.registerColumn(translate("id"), "id")
      sheet.registerColumn(translate("created"), "created").withSize(ExcelUtils.Size.TIMESTAMP)
      sheet.registerColumn(translate("lastUpdate"), "lastUpdate").withSize(ExcelUtils.Size.TIMESTAMP)
      ExcelUtils.addHeadRow(sheet, boldStyle)

      list.forEach { timesheet ->
        val row = sheet.createRow()
        row.autoFillFromObject(timesheet)
        val node = taskTree.getTaskNodeById(timesheet.taskId)
        val user = userGroupCache.getUser(timesheet.userId)
        row.getCell("user")?.setCellValue(user?.getFullname())
        row.getCell("kunde")?.setCellValue(timesheet.kost2?.projekt?.kunde?.name)
        row.getCell("projekt")?.setCellValue(timesheet.kost2?.projekt?.name)
        row.getCell("kost2")?.setCellValue(timesheet.kost2?.displayName)
        row.getCell("task.title")?.setCellValue(node.task.title)
        row.getCell("taskPath")?.setCellValue(getTaskPath(timesheet.taskId, null, true, OutputType.PLAIN))
        row.getCell("weekOfYearShortLabel")?.setCellValue(timesheet.getFormattedWeekOfYear())
        row.getCell("dayOfWeekShortLabel")?.setCellValue(
          dateTimeFormatter.getFormattedDate(
            timesheet.startTime, getFormatString(DateFormatType.DAY_OF_WEEK_SHORT)
          )
        )
        row.getCell(stopTimeColDef).cell.cellStyle = timeFormat
        // ExcelUtils.getCell(row, TimesheetDO::startTime)?.setCellValue(startTime)
        // ExcelUtils.getCell(row, TimesheetDO::stopTime)?.setCellValue(stopTime)
        val seconds = BigDecimal(timesheet.getDuration() / 1000) // Seconds
        val duration = seconds.divide(BigDecimal(60 * 60 * 24), 8, RoundingMode.HALF_UP) // Fraction of day (24 hours)
        row.getCell("duration")?.setCellValue(duration.toDouble())?.setCellStyle(durationFormat)
        val hours = seconds.divide(BigDecimal(60 * 60), 2, RoundingMode.HALF_UP)
        row.getCell("hours")?.setCellValue(hours.toDouble())?.setCellStyle(hoursFormat)
        row.getCell("taskReference")?.setCellValue(node.reference)
        row.getCell("shortDescription")?.setCellValue(timesheet.getShortDescription())?.setCellStyle(wrapTextStyle)
        row.getCell(descriptionColDef).cell.cellStyle = wrapTextStyle
        row.getCell("created")?.setCellValue(timesheet.created)
        row.getCell("lastUpdate")?.setCellValue(timesheet.lastUpdate)
        ExcelUtils.getCell(row, TimesheetDO::description)?.setCellStyle(wrapTextStyle)
      }
      sheet.setAutoFilter()
      sheet.createFreezePane(stopTimeColDef.columnNumber + 1, 1)
      return workbook.asByteArrayOutputStream.toByteArray()
    }
  }
}
